std/array.d

```diff
@@ -3927,3 +3927,220 @@ unittest
+}
+
+/// static array from array
+nothrow pure @safe unittest
+{
+    auto a = [0, 1].asStatic;
+    static assert(is(typeof(a) == int[2]));
+    assert(a == [0, 1]);
+
+    auto b = [0, 1].asStatic!byte;
+    static assert(is(typeof(b) == byte[2]));
+    assert(b == [0, 1]);
+}
+
+nothrow pure @safe unittest
+{
+    int val = 3;
+    static immutable gold = [1, 2, 3];
+    [1, 2, val].asStatic.checkStaticArray!int([1, 2, 3]);
+
+    @nogc void checkNogc()
+    {
+        [1, 2, val].asStatic.checkStaticArray!int(gold);
+    }
+    checkNogc();
+
+    [1, 2, val].asStatic!double.checkStaticArray!double(gold);
+    [1, 2, 3].asStatic!int.checkStaticArray!int(gold);
+
+    [1, 2, 3].asStatic!(const(int)).checkStaticArray!(const(int))(gold);
+    [1, 2, 3].asStatic!(const(double)).checkStaticArray!(const(double))(gold);
+    {
+        const(int)[3] a2 = [1, 2, 3].asStatic;
+    }
+
+    [1, 129].asStatic!byte.checkStaticArray!byte([1, -127]);
+
+}
+
+/// ditto
+auto asStatic(size_t n, T)(T a)
+{
+    import std.conv : emplaceRef;
+    // TODO: ElementType vs ForeachType
+    alias U = typeof(a[0]);
+    U[n] ret = void;
+    size_t i;
+    foreach (ref ai; a)
+    {
+        emplaceRef!U(ret[i++], ai);
+    }
+    assert(i == n);
+    return ret;
+}
+
+/// ditto
+auto asStatic(Un : U[n], U, size_t n, T)(T a) nothrow @safe pure @nogc
+{
+    import std.conv : emplaceRef;
+    U[n] ret = void;
+    size_t i;
+    foreach (ref ai; a)
+    {
+        emplaceRef!U(ret[i++], cast(U) ai);
+    }
+    assert(i == n);
+    return ret;
+}
+
+/// static array from range + size
+nothrow pure @safe unittest
+{
+    import std.range : iota;
+    auto input = 2.iota;
+    auto a = input.asStatic!2;
+    static assert(is(typeof(a) == int[2]));
+    assert(a == [0, 1]);
+    auto b = input.asStatic!(byte[2]);
+    static assert(is(typeof(b) == byte[2]));
+    assert(b == [0, 1]);
+}
+
+nothrow pure @safe unittest
+{
+
+    auto a = [1, 2].asStatic;
+    assert(is(typeof(a) == int[2]) && a == [1, 2]);
+
+    import std.range : iota;
+
+    2.iota.asStatic!2.checkStaticArray!int([0, 1]);
+    2.iota.asStatic!(double[2]).checkStaticArray!double([0, 1]);
+    2.iota.asStatic!(byte[2]).checkStaticArray!byte([0, 1]);
+}
+
+nothrow pure @system unittest
+{
+    import std.range : iota;
+
+    assert(isThrown!Error(2.iota.asStatic!1));
+    assert(isThrown!Error(2.iota.asStatic!3));
+    // NOTE: alternatives are not nothrow:
+    version (none)
+    {
+        import std.exception : assertThrown;
+
+        assertThrown!Error(2.iota.asStatic!3);
+        import std.exception : ifThrown;
+
+        assert(ifThrown!Error({ 2.iota.asStatic!1; return false; }(), true));
+    }
+}
+
+@system unittest
+{
+    version (Bug)
+    {
+        // https://issues.dlang.org/show_bug.cgi?id=16779
+        auto a2 = [1, 2, 3].asStatic!byte;
+        auto a3 = [1, 2, 3].asStatic!ubyte;
+
+        // NOTE: correctly issues a deprecation
+        int[] a2 = [1, 2].asStatic;
+    }
+}
```

Comment:
Not sure whether it's a good idea to include such dead blocks. Maybe open a Bugzilla issue to fix asStatic that depends on 16779 so that this isn't forgotten?

Reply:
actually 16779 was for VRP on staticArray(T...)(T a) which I removed (i just have asStatic now), so removed that;
also, folded other dead block into a comment in preceding unittest.

__traits(deprecated) would be the optimal (see https://dlang.slack.com/archives/D8ZMKRGUV/p1519414575000349)